compiler/rustc_typeck/src/check/callee.rs

```diff
@@ -307,6 +307,37 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    /// Give appropriate suggestion when encountering `[("a", 0) ("b", 1)]`, where the
+    /// likely intention is to create an array containing tuples.
+    fn identify_bad_array_def(
+        &self,
+        err: &mut DiagnosticBuilder<'a>,
+        call_expr: &'tcx hir::Expr<'tcx>,
+        callee_node: &hir::ExprKind<'_>,
+        callee_span: Span,
+    ) -> bool {
+        let hir_id = self.tcx.hir().get_parent_node(call_expr.hir_id);
+        let parent_node = self.tcx.hir().get(hir_id);
+        if let (
+            hir::Node::Expr(hir::Expr { kind: hir::ExprKind::Array(_), .. }),
+            hir::ExprKind::Tup(exp),
+            hir::ExprKind::Call(_, args),
+        ) = (parent_node, callee_node, &call_expr.kind)
+        {
+            if args.len() == exp.len() {
+                let start = callee_span.shrink_to_hi();
+                err.span_suggestion(
+                    start,
+                    "try adding this here",
```

Comment:
```suggestion
                    "consider separating array elements with a comma",
```
I think this would be better if we described the contents of the suggestion more.